master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,597 @@
+
+# Duplicated from determined/__init__.py.  It's nice to keep them in sync.
+LOG_FORMAT = "%(levelname)s: [%(process)s] %(name)s: %(message)s"
+
+
+# Example log message given below.
+# 2022-05-12 16:32:48,757:gc_checkpoints: [rank=0] INFO: Determined checkpoint GC, ...
+# Below regex is used to extract the rank field from the log message.
+# Excluding empty spaces this regex matches rank in the above example as [rank=0]
+rank = re.compile(r"(?P<space1> ?)\[rank=(?P<rank_id>([0-9]+))\](?P<space2> ?)(?P<log>.*)")
+# Below regex is used to extract the message severity from the log message.
+# Excluding empty spaces and delimiter(:) this regex matches message severity level in the above
+# example as INFO
+level = re.compile(
+    r"(?P<space1> ?)(?P<level>(DEBUG|INFO|WARNING|ERROR|CRITICAL)):(?P<space2> ?)(?P<log>.*)"
+)
+lineend = re.compile(rb"[\r\n]")
+
+
+# Interval at which to force a flush.
+SHIPPER_FLUSH_INTERVAL = 1  # How often to make API calls
+
+# Full jitter time on encountering an API exception.
+SHIPPER_FAILURE_BACKOFF_SECONDS = 1
+
+# Max size of the log buffer before forcing a flush.
+LOG_BATCH_MAX_SIZE = 1000
+
+# Max size of the shipping queue before we start to apply backpressure by blocking sends. We would
+# only hit this if we got underwater by three full batches while trying to ship a batch.
+SHIP_QUEUE_MAX_SIZE = 3 * LOG_BATCH_MAX_SIZE
+
+
+def read_newlines_or_carriage_returns(fd: IO[bytes]) -> Iterator[str]:
+    r"""
+    Read lines, delineated by either '\n' or '\r.
+
+    Unlike the default io.BufferedReader used in subprocess.Popen(bufsize=-1), we read until we
+    encounter either '\n' or \r', and treat that as one line.
+
+    Specifically, io.BufferedReader doesn't handle tqdm progress bar outputs very well; it treats
+    all of the '\r' outputs as one enormous line.
+
+    Args:
+        fd: an unbuffered stdout or stderr from a subprocess.Popen.
+
+    Yields:
+        A series of str, one per line.  Each line always ends with a '\n'.
+    """
+    # Ship lines of length of DEFAULT_BUFFER_SIZE, including the terminating newline.
+    limit = io.DEFAULT_BUFFER_SIZE - 1
+    nread = 0
+    chunks = []  # type: List[bytes]
+
+    def oneline():
+        nonlocal nread
+        nonlocal chunks
+        out = b"".join(chunks).decode("utf8")
+        chunks = []
+        nread = 0
+        return out
+
+    while True:
+        buf = fd.read(limit - nread)
+        if not buf:
+            # Normal EOF.
+            break
+
+        # Extract all the lines from this buffer.
+        while buf:
+            m = lineend.search(buf)
+            if m is None:
+                # No line break here; just append to chunks.
+                chunks.append(buf)
+                nread += len(buf)
+                break
+
+            # Line break found!
+            start, end = m.span()
+            chunks.append(buf[:start])
+            # Even if we matched a '\r', emit a '\n'.
+            chunks.append(b"\n")
```

Comment:
what you have is equivalent to current behavior but i had a thought to handle `\r\n` as a single `\n`

Reply:
Handling `\r\n` as a single `\n` when processing a stream is like one the world's most annoying problems, because what if somebody is sending `\r`-terminated lines?  Do you wait for the next byte?  Or do you call this `\r` a complete line right now?

In practice, since this code only runs on linux anyway, it's not really worth trying to solve that problem, since `\n`-terminated lines are universal on linux.